### ISSUE TITLE
Array and pattern matching

### ISSUE THREAD

**Issue by Namek** [mint-lang/mint#225]
**TL;DR** 1 documentation bug, 1 compiler bug

------

[Documentation](https://www.mint-lang.com/guide/reference/arrays) says:

```
  [head, ...tail] => ""   /* Array at least two elements */
```

The comment tells that `tail` cannot be empty for this case to match. But I experience something different (which I actually prefer).

Let's have a simple program:

```
component Main {
  fun render : String {
    try {
      arr =
        ["1"]

      transform(arr)
    }
  }

  fun transform (arr : Array(String)) : String {
    case (arr) {
      [x, ...rest] => x
      => "0"
    }
  }
}

```

For the program above the complier errors with:
```
All possibilities of the case expression are covered. 

This branch is not needed and can be safely removed. 


┌──┬ ./Main.mint ──────────────────────────────────────────────────────────────┐
│11│                                                                           │
│12│   fun transform(arr : Array(String)) : String {                           │
│13│   case (arr) {                                                            │
│14│         [x, ...rest] => x                                                 │
│15│         => "0"                                                            │
│16│       }                                                                   │
│17│   }                                                                       │
│18│ }                                                                         │
└──┴───────────────────────────────────────────────────────────────────────────┘
```

Let's delete the `=> "0"` case then. Then it compiles and prints "1". Which is what I would expect but the comment from the documentation lies becauses the array consists of only 1 element, not two.

Similar with `[...arr]` case. I would expect at least 0 elements.

https://sandbox.mint-lang.com/sandboxes/aTlMykaTP2kOEQ

**Comment by gdotdesign** [comment]
Yes the documentation is wrong any array destructuring with a spread (`...rest`) does not have any size condition.

What I don't see is the compiler bug, so if the documentation is wrong then there is no compiler bug?

**Comment by Namek** [comment]
The compiler bug is another issue, please read the code snippet and the error. It tells me all possibilities are covered without the default case which is not true.

**Comment by gdotdesign** [comment]
Yeah I get it now sorry.

**Comment by gdotdesign** [comment]
I've updated the algorithm that checks for the coverage, not it takes into account the empty literal as well as the number of items each destructuring has.

**Comment by Namek** [comment]
Yeah it seems to work in the case I mentioned. Now only the documentation is left.

**Comment by Namek** [comment]
@gdotdesign Well, sorry, no, it doesn't work properly.

The `[x, ...rest]` matches when collection has 0 items.

```
try {
  arr =
    []

  case (arr) {
    [head, ...tail] => "false"
    => "true"
  }
}
```

`"false"` is the result.

**Comment by gdotdesign** [comment]
You are right I've not updated the compiler accordingly.

**Comment by gdotdesign** [comment]
It should be fixed with this 3d8cbcaf824ed6f7012248dc81e3f0ea154330b7

**Comment by Namek** [comment]
@gdotdesign more issues:

### No case matched dynamically when array size is 1

The function below returns `undefined`:

```
fun test {
  try {
    arr =
      [2]

    case (arr) {
      [] => "lol0"

      [head, ...tail] =>
        try {
          "lol1"
        }
    }
  }
}
```

Basically it does not enter into any case. If the input array is empty then it falls into `lol0` of course.

### Syntax error with `try`

Another issue:

```
    case (arr) {
      [] => try { "lol0" }

      [head, ...tail] =>
        try {
          "lol1"
        }
    }
```

![image](https://user-images.githubusercontent.com/31058/78512227-0249ff80-77a3-11ea-9e0b-e3828301253f.png)

**Comment by gdotdesign** [comment]
> No case matched dynamically when array size is 1

This is fixed by b313f015828eeb5f0bb6c87698a6384df5dabe57

> Syntax error with try

This is harder to fix, basically it parses the next case branch as an array access `[0]`, I need to think about how to handle this particular case.

**Comment by gdotdesign** [comment]
Created a new issue for the syntax error, so we can close this one if fixed.

**Comment by Namek** [comment]
Indeed, it works for me now. Howvever, the documentation is still not updated. @gdotdesign

**Comment by gdotdesign** [comment]
Updated the documentation in https://github.com/mint-lang/mint-website-rails/commit/edcf9092aa10085e290de030bd08d566b704c857
